homeassistant/components/decora_wifi/__init__.py

```diff
@@ -1 +1,129 @@
-"""The decora_wifi component."""
+"""Support for the myLeviton decora_wifi component."""
+
+import logging
+
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.const import (
+    CONF_HOST,
+    CONF_PASSWORD,
+    CONF_SCAN_INTERVAL,
+    CONF_USERNAME,
+)
+from homeassistant.core import HomeAssistant
+import homeassistant.helpers.config_validation as cv
+
+from .common import (
+    DecoraWifiCommFailed,
+    DecoraWifiLoginFailed,
+    DecoraWifiPlatform,
+    decorawifisessions,
+)
+from .const import CONF_OPTIONS, DEFAULT_SCAN_INTERVAL, DOMAIN, PLATFORMS
+
+_LOGGER = logging.getLogger(__name__)
+
+DECORAWIFI_HOST_SCHEMA = vol.Schema({vol.Required(CONF_HOST): cv.string})
+
+CONFIG_SCHEMA = vol.Schema(
+    vol.All(
+        cv.deprecated(DOMAIN),
+        {
+            DOMAIN: vol.Schema(
+                {
+                    vol.Required(CONF_USERNAME): cv.string,
+                    vol.Required(CONF_PASSWORD): cv.string,
+                    vol.Optional(CONF_SCAN_INTERVAL, default=120): cv.positive_int,
+                },
+            )
+        },
+    ),
+    extra=vol.ALLOW_EXTRA,
+)
+
+
+async def async_setup(hass, config):
+    """Set up the Decora Wifi component."""
+    hass.data.setdefault(DOMAIN, {})
+
+    if DOMAIN not in config:
+        return True
+
+    # Trigger import config flow
+    conf = config[DOMAIN]
+
+    if not hass.config_entries.async_entries(DOMAIN):
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN,
+                context={"source": config_entries.SOURCE_IMPORT},
+                data={
+                    CONF_USERNAME: conf[CONF_USERNAME],
+                    CONF_PASSWORD: conf[CONF_PASSWORD],
+                },
+            )
+        )
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry):
+    """Set up Decora Wifi from a config entry."""
+
+    conf_data = dict(entry.data)
+    email = conf_data[CONF_USERNAME]
+    password = conf_data[CONF_PASSWORD]
+
+    conf_data[CONF_OPTIONS] = {}
+    # If no scan interval option was set in the config entry, use the default.
+    conf_data[CONF_OPTIONS].update(
+        {
+            CONF_SCAN_INTERVAL: dict(entry.options).get(
+                CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL
+            )
+        }
+    )
+
+    # If a session is already setup, don't start a new one.
+    session = decorawifisessions.get(email, None)
```

Comment:
Store `decorawifisessions` under `hass.data[DOMAIN]` so you can clean them up when the integration is unloaded or deleted

Reply:
I did try that earlier on in the development process and agree it would be better if I could get it to work, but it threw an error message basically saying that the upstream API's session object is not serializable.  Sorry, I don't have the exact error message in front of me.

Reply:
Managed to reproduce the message in my test environment this morning:
```
Error writing config for core.config_entries: Failed to serialize to JSON: /config/.storage/core.config_entries. Bad data at $.data.entries[9].data.session=<custom_components.decora_wifi.common.DecoraWifiPlatform object at 0x7ff3385796a0>(<class 'custom_components.decora_wifi.common.DecoraWifiPlatform'>
```

Reply:
Looks like most other integrations just store an auth token in hass.data, which is not something this one is really designed to do.

I'm thinking the answer is to put the session object in an entity (maybe using a BinarySensor and having it indicate if the API becomes unavailable) and add that entity to hass during `__init__` and/or `config flow`. Then, during each platform's `async_setup_entry` we can retrieve that Entity from the entity registry and retrieve a reference to the session that way.  All the platforms end up with access to the same session object, but the session object stays correctly tied to the hass lifecycle functions so it gets cleaned up when the integration is unloaded or deleted.

Thoughts?